src/sentry/grouping/enhancer/__init__.py

```diff
@@ -228,6 +237,85 @@ def compare_rust_enhancers(
                 sentry_sdk.capture_message("Rust Enhancements mismatch")
 
 
+def assemble_rust_components(
+    rust_enhancements: RustEnhancements | None,
+    match_frames: list[dict[str, bytes]],
+    exception_data: dict[str, Any],
+    components: list[GroupingComponent],
+) -> RustAssembleResult | None:
+    """
+    If `RustEnhancements` were successfully parsed and usage is enabled,
+    this will update all the frame `components` contributions.
+
+    This primarily means updating the `contributes`, `hint` as well as other attributes
+    of each frames `GroupingComponent`.
+    Instead of modifying the input `components` directly, the results are returned
+    as a list of `RustComponent`.
+    """
+    if not rust_enhancements:
+        return None
+
+    try:
+        rust_components = [
+            RustComponent(
+                is_prefix_frame=c.is_prefix_frame or False,
+                is_sentinel_frame=c.is_sentinel_frame or False,
+                contributes=c.contributes,
+            )
+            for c in components
+        ]
+
+        rust_results = rust_enhancements.assemble_stacktrace_component(
+            match_frames, make_rust_exception_data(exception_data), rust_components
+        )
+
+        return (
+            rust_results.contributes,
+            rust_results.hint,
+            rust_results.invert_stacktrace,
+            rust_components,
+        )
+    except Exception:
+        logger.exception("failed running Rust Enhancements component contributions")
+        return None
+
+
+def compare_rust_components(
+    component: GroupingComponent,
+    invert_stacktrace: bool,
+    rust_results: RustAssembleResult | None,
+    frames: Sequence[dict[str, Any]],
+):
+    """
+    Compares the results of `rust_results` with the component modifications
+    applied by Python code directly to `components`.
+
+    This will log an internal error on every mismatch.
+    """
+    if not rust_results:
+        return
+
+    contributes, hint, invert, rust_components_ = rust_results
+
+    python_components = [
+        (c.contributes, c.is_prefix_frame, c.is_sentinel_frame) for c in component.values
+    ]
+    rust_components = [
+        (c.contributes, c.is_prefix_frame, c.is_sentinel_frame) for c in rust_components_
+    ]
+
+    python_res = (component.contributes, component.hint, invert_stacktrace, python_components)
+    rust_res = (contributes, hint, invert, rust_components)
+
+    if python_res != rust_res:
```

Comment:
You should probably have test coverage to make sure that the comparison works as you expect.
I'm okay with you merging as-is but we need to make sure that the reporting will work as expected.

Reply:
TBH, I don’t ever expect this code to be hit. If it is, it means our implementation is buggy and does not match the expectation.
So its not quite possible to create a test to cover it, because that would reveal a bug which should then rather be fixed ;-)